File: scripts/trade_rules.py
"""Evaluate reviewed Pokémon GO trade rules from a versioned local registry."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Mapping

REGISTRY_PATH = Path(__file__).resolve().parents[1] / "knowledge" / "trade-rules.v1.json"


def load_registry(path: Path | None = None) -> dict[str, Any]:
    return json.loads((path or REGISTRY_PATH).read_text(encoding="utf-8"))


def _known_bool(values: Mapping[str, Any], key: str) -> bool | None:
    value = values.get(key)
    return value if isinstance(value, bool) else None


def evaluate_trade(
    mode: str,
    pokemon: Mapping[str, Any],
    friend: Mapping[str, Any] | None = None,
    trainer: Mapping[str, Any] | None = None,
    registry: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    rules = dict(registry or load_registry())
    mode_rules = (rules.get("modes") or {}).get(mode)
    if mode_rules is None:
        raise ValueError(f"Unsupported trade mode: {mode}")

    friend = dict(friend or {})
    trainer = dict(trainer or {})
    blockers: list[str] = []
    unknowns: list[str] = []

    minimum = mode_rules.get("trainer_min_level")
    if minimum is not None:
        level = trainer.get("level")
        if level is None:
            unknowns.append("trainer_level")
        elif int(level) < int(minimum):
            blockers.append("trainer_level_below_minimum")

    if mode_rules.get("requires_forever_friend"):
        value = _known_bool(friend, "forever_friend")
        if value is None:
            unknowns.append("forever_friend")
        elif value is False:
            blockers.append("not_forever_friends")

    if mode_rules.get("requires_available_remote_trade"):
        value = _known_bool(friend, "remote_trade_available")
        if value is None:
            unknowns.append("remote_trade_available")
        elif value is False:
            blockers.append("no_remote_trade_available")

    if mode == "remote":
        count = friend.get("remote_trades_completed_today")
        if count is None:
            unknowns.append("remote_trades_completed_today")
        elif int(count) >= int(rules["friendship"]["remote_trade"]["completed_per_day_limit"]):
            blockers.append("remote_daily_limit_reached")

    for key in mode_rules.get("hard_blockers") or []:
        value = _known_bool(pokemon, key)
        if value is None:
            unknowns.append(key)
        elif value:
            blockers.append(key)

    state = "blocked" if blockers else "unknown" if unknowns else "eligible"
    special = False if mode == "remote" else any(
        _known_bool(pokemon, key) is True for key in mode_rules.get("special_trade_categories") or []
    )

    return {
        "state": state,
        "mode": mode,
        "blockers": sorted(set(blockers)),
        "unknowns": sorted(set(unknowns)),
        "special_trade": special,
        "lucky": _known_bool(friend, "lucky_friend") is True,
        "post_trade_stats_guaranteed": False,
        "exact_stardust_cost": None,
        "requires_game_confirmation": True,
        "reviewed_at": rules.get("reviewed_at"),
    }


__all__ = ["REGISTRY_PATH", "load_registry", "evaluate_trade"]

File: site/trade-rules.js
"use strict";

(function exposeTradeRules(root, factory) {
  const api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  if (root) root.CollectionTradeRules = api;
})(typeof globalThis !== "undefined" ? globalThis : this, () => {
  const knownBool = (values, key) => typeof values?.[key] === "boolean" ? values[key] : null;
  const knownNumber = (value) => {
    if (value === undefined || value === null || value === "") return null;
    const number = Number(value);
    return Number.isFinite(number) ? number : null;
  };
  const uniqueSorted = (values) => [...new Set(values)].sort();

  function evaluateTrade(mode, pokemon = {}, friend = {}, trainer = {}, registry = {}) {
    const modeRules = registry?.modes?.[mode];
    if (!modeRules) throw new Error(`Unsupported trade mode: ${mode}`);

    const blockers = [];
    const unknowns = [];
    const minimum = modeRules.trainer_min_level;
    if (minimum !== undefined && minimum !== null) {
      const level = knownNumber(trainer?.level);
      if (level === null) unknowns.push("trainer_level");
      else if (level < Number(minimum)) blockers.push("trainer_level_below_minimum");
    }

    if (modeRules.requires_forever_friend) {
      const value = knownBool(friend, "forever_friend");
      if (value === null) unknowns.push("forever_friend");
      else if (!value) blockers.push("not_forever_friends");
    }

    if (modeRules.requires_available_remote_trade) {
      const value = knownBool(friend, "remote_trade_available");
      if (value === null) unknowns.push("remote_trade_available");
      else if (!value) blockers.push("no_remote_trade_available");
    }

    if (mode === "remote") {
      const count = knownNumber(friend?.remote_trades_completed_today);
      if (count === null) unknowns.push("remote_trades_completed_today");
      else if (count >= Number(registry?.friendship?.remote_trade?.completed_per_day_limit ?? 1)) blockers.push("remote_daily_limit_reached");
    }

    for (const key of modeRules.hard_blockers || []) {
      const value = knownBool(pokemon, key);
      if (value === null) unknowns.push(key);
      else if (value) blockers.push(key);
    }

    const normalizedBlockers = uniqueSorted(blockers);
    const normalizedUnknowns = uniqueSorted(unknowns);
    const state = normalizedBlockers.length ? "blocked" : normalizedUnknowns.length ? "unknown" : "eligible";
    const specialTrade = mode === "remote" ? false : (modeRules.special_trade_categories || []).some((key) => knownBool(pokemon, key) === true);

    return {
      state,
      mode,
      blockers: normalizedBlockers,
      unknowns: normalizedUnknowns,
      special_trade: specialTrade,
      lucky: knownBool(friend, "lucky_friend") === true,
      post_trade_stats_guaranteed: false,
      exact_stardust_cost: null,
      requires_game_confirmation: true,
      reviewed_at: registry?.reviewed_at || null,
    };
  }

  function summarizeEligibility(evaluation) {
    if (!evaluation || !evaluation.state) return "Trade eligibility is unavailable.";
    if (evaluation.state === "blocked") return `Blocked: ${evaluation.blockers.join(", ")}.`;
    if (evaluation.state === "unknown") return `Needs confirmation: ${evaluation.unknowns.join(", ")}.`;
    return `Eligible under reviewed ${evaluation.mode === "remote" ? "Remote Trade" : "in-person trade"} rules; confirm final cost and game state in Pokémon GO.`;
  }

  return { evaluateTrade, summarizeEligibility };
});
